src/backend/distributed/planner/distributed_planner.c

```diff
@@ -2534,3 +2556,112 @@ WarnIfListHasForeignDistributedTable(List *rangeTableList)
 		}
 	}
 }
+
+
+/*
+ * AreTablesMergeSqlCommandSupported returns true if the tables in the MERGE
+ * command are supported, the two scenario supported are
+ * -- All the tables are Citus local
+ * -- All the tables are distributed and co-located
+ * Note: We are deferring the expensive check of whether MERGE joined on the
+ *       distribution column to the push-down planning phase.
+ */
+static bool
+AreTablesMergeSqlCommandSupported(List *rangeTableList)
+{
+	List *localList = NIL;
+	List *distList = NIL;
+	ListCell *tableCell = NULL;
+
+	foreach(tableCell, rangeTableList)
+	{
+		RangeTblEntry *rangeTableEntry = (RangeTblEntry *) lfirst(tableCell);
+		Oid relationId = rangeTableEntry->relid;
+
+		/* Skip CTEs, sub-queries */
+		if (rangeTableEntry->rtekind == RTE_CTE ||
+			rangeTableEntry->rtekind == RTE_SUBQUERY)
+		{
+			continue;
+		}
+
+		/* Reference tables are not supported */
+		if (IsCitusTableType(relationId, REFERENCE_TABLE))
+		{
+			ereport(ERROR,
+					(errcode(ERRCODE_FEATURE_NOT_SUPPORTED),
+					 errmsg("MERGE command is not supported on Reference tables yet")));
+		}
+
+		/* Combination of Postgres and Citus tables is not supported yet */
+		if (!IsCitusTable(relationId))
```

Comment:
I think we should be able to support subqueries in the source that are also joined in dist keys. The example is non-sense, but the point holds: 

```SQL
MERGE INTO target t                            USING (SELECT source.* FROM source JOIN target ON (sid=tid)) AS s                        ON t.tid = s.sid                                                                         WHEN MATCHED THEN                                                                        UPDATE SET tid = 0;
ERROR:  MERGE command is not supported with combination of Postgres and Citus tables
Time: 0.428 ms
```

The thing that I'm mostly curious is that whether all the restriction information (e.g., all the joins and filters that PG keeps track of) that Citus planners heavily relies on works fine in MERGE command.

Reply:
I've added a test with local tables 

-- Joining on columns inside the sub-query
MERGE INTO t3
        USING (SELECT s3_1.id, s3_2.val, s3_2.src FROM s3_1, s3_2 WHERE s3_1.id = s3_2.id) sub
        ON (t3.id = sub.id)
        WHEN MATCHED AND sub.val = 1 THEN
                UPDATE SET val = t3.val + 1
        WHEN MATCHED THEN
                DELETE
        WHEN NOT MATCHED THEN
                INSERT (id, val, src) VALUES (sub.id, sub.val, sub.src);